### PR TITLE
Migrate Compression API to new Buffer API

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We introduced a new Buffer API in netty 5 but didnt migrate the Compression
to use it yet

Modifications:

- Port Compression API and implementations to use new Buffer API
- Adjust tests

Result:

Use new Buffer API in compression implementations

### CODE REVIEW

**Comment by normanmaurer** [comment]
This also includes https://github.com/netty/netty/pull/12424 which I will remove from this PR once https://github.com/netty/netty/pull/12424 is merged

File: codec/src/test/java/io/netty5/handler/codec/compression/SnappyFrameEncoderTest.java [hunk 1]
```diff
@@ -35,76 +34,54 @@ public void setUp() {
...
-        });
-
-        CompositeByteBuf actual = Unpooled.compositeBuffer();
-        for (;;) {
-            ByteBuf m = channel.readOutbound();
-            if (m == null) {
-                break;
-            }
-            actual.addComponent(true, m);
+        }); CompositeBuffer actual = CompressionTestUtils.compose(channel.bufferAllocator(), channel::readOutbound)) {
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/SnappyFrameEncoderTest.java:82]: Missing line break?

File: codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java [hunk 2]
```diff
@@ -39,304 +40,269 @@ public void resetSnappy() {
...
-        });
-        assertEquals(expected, out, "Encoded literal was invalid");
-
-        in.release();
-        out.release();
-        expected.release();
+    public void encodeShortTextIsLiteral() {
+        try (Buffer in =  BufferAllocator.onHeapUnpooled().copyOf(new byte[] {
+                0x6e, 0x65, 0x74, 0x74, 0x79
+        }); Buffer out = BufferAllocator.onHeapUnpooled().allocate(7);
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java:125]: Missing line break?

Comment (chrisvest) [reply codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java:125]: Oh, we're still in the resource expression here. Still looks weird, though.

File: codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java [hunk 3]
```diff
@@ -39,304 +40,269 @@ public void resetSnappy() {
...
-        assertEquals(CharBuffer.wrap(srcStr),
-            CharBuffer.wrap(outDecoded.getCharSequence(0, outDecoded.writerIndex(), CharsetUtil.US_ASCII)));
-
-        in.release();
-        out.release();
-        outDecoded.release();
+                // copy of "rotocols"
+                0x11, 0x4c,
+        };
+        try (Buffer in = BufferAllocator.onHeapUnpooled().copyOf(srcStr.getBytes(StandardCharsets.US_ASCII));
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java:197]:
```suggestion
        try (Buffer in = BufferAllocator.onHeapUnpooled().copyOf(srcStr, StandardCharsets.US_ASCII);
```

File: codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java [hunk 4]
```diff
@@ -39,304 +40,269 @@ public void resetSnappy() {
...
+            Buffer out = BufferAllocator.onHeapUnpooled().allocate(180);
+            Buffer outDecoded = BufferAllocator.onHeapUnpooled().allocate(256);
+            Buffer expected = BufferAllocator.onHeapUnpooled().copyOf(expectedBytes);
+            Buffer expectedDecoded = BufferAllocator.onHeapUnpooled().copyOf(
+                    srcStr.getBytes(StandardCharsets.US_ASCII))) {
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java:202]:
```suggestion
            Buffer expectedDecoded = BufferAllocator.onHeapUnpooled().copyOf(
                    srcStr, StandardCharsets.US_ASCII)) {
```

File: codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java [hunk 5]
```diff
@@ -39,304 +40,269 @@ public void resetSnappy() {
...
 
     @Test
     public void testLarge2ByteLiteralLengthAndCopyOffset() {
-        ByteBuf compressed = Unpooled.buffer();
-        ByteBuf actualDecompressed = Unpooled.buffer();
-        ByteBuf expectedDecompressed = Unpooled.buffer().writeByte(0x01).writeZero(0x8000).writeByte(0x01);
-        try {
+        try (Buffer compressed = BufferAllocator.onHeapUnpooled().allocate(256);
+            Buffer actualDecompressed = BufferAllocator.onHeapUnpooled().allocate(256);
+             Buffer expectedDecompressed = BufferAllocator.onHeapUnpooled().allocate(256)) {
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/SnappyTest.java:275]: Wonky indentation.

File: codec/src/test/java/io/netty5/handler/codec/compression/ZstdEncoderTest.java [hunk 6]
```diff
@@ -52,34 +51,32 @@ public EmbeddedChannel createChannel() {
...
         final int dataLength = data.readableBytes();
-        int written = 0;
+        try (Buffer expected = data.copy()) {
+            Buffer in = data.readSplit(65535);
+            assertTrue(channel.writeOutbound(in));
 
-        ByteBuf in = data.retainedSlice(written, 65535);
-        assertTrue(channel.writeOutbound(in));
+            Buffer in2 = data.readSplit(dataLength - 65535);
+            assertTrue(channel.writeOutbound(in2));
```
Comment (chrisvest) [inline codec/src/test/java/io/netty5/handler/codec/compression/ZstdEncoderTest.java:61]: `in2` would contain all the remaining readable bytes of `data`, so can't we just pass `data` here?
If so, then we could also inline `in` above, so neither local variable is necessary.

File: codec/src/main/java/io/netty5/handler/codec/compression/ByteBufChecksum.java [hunk 7]
```diff
@@ -15,110 +15,66 @@
...
 
 /**
- * {@link Checksum} implementation which can directly act on a {@link ByteBuf}.
- *
- * Implementations may optimize access patterns depending on if the {@link ByteBuf} is backed by a
- * byte array ({@link ByteBuf#hasArray()} is {@code true}) or not.
+ * {@link Checksum} implementation which can directly act on a {@link Buffer}.
  */
-abstract class ByteBufChecksum implements Checksum {
+class ByteBufChecksum implements Checksum {
```
Comment (chrisvest) [inline codec/src/main/java/io/netty5/handler/codec/compression/ByteBufChecksum.java:28]: We could rename this to `BufferChecksum`

**Review by chrisvest — CHANGES_REQUESTED** [review]
Looks good overall. Had a number of small comments.

File: codec/src/main/java/io/netty5/handler/codec/compression/BrotliCompressor.java [hunk 8]
```diff
@@ -86,21 +84,22 @@ public static Supplier<BrotliCompressor> newFactory() {
...
+                return allocator.allocate(0);
             case PROCESSING:
-                if (!input.isReadable()) {
-                    return Unpooled.EMPTY_BUFFER;
+                if (input.readableBytes() == 0) {
+                    return allocator.allocate(0);
                 }
-                byte[] uncompressed = ByteBufUtil.getBytes(input, input.readerIndex(), input.readableBytes(), false);
+                // TODO: Improve me
+                byte[] uncompressed = new byte[input.readableBytes()];
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/BrotliCompressor.java:98]: future is fine imho

File: codec/src/main/java/io/netty5/handler/codec/compression/ZstdCompressor.java [hunk 9]
```diff
@@ -129,26 +128,27 @@ private ByteBuf allocateBuffer(ByteBufAllocator allocator, ByteBuf msg) {
                     "the maximum allowable size (" + maxEncodeSize + " bytes)");
         }
 
-        return msg.isDirect() ? allocator.directBuffer((int) bufferSize) : allocator.heapBuffer((int) bufferSize);
+        // TODO: It would be better if we could allocate depending on the input type
+        return allocator.allocate((int) bufferSize);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/ZstdCompressor.java:132]: yes and and also if an allocator would be able to allocate direct and heap buffers just like it did in 4.1

File: codec/src/test/java/io/netty5/handler/codec/compression/AbstractDecoderTest.java [hunk 10]
```diff
@@ -31,11 +30,6 @@
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
 public abstract class AbstractDecoderTest extends AbstractCompressionTest {
 
-    protected static final ByteBuf WRAPPED_BYTES_SMALL = Unpooled.unreleasableBuffer(
-            Unpooled.wrappedBuffer(BYTES_SMALL)).asReadOnly();
-    protected static final ByteBuf WRAPPED_BYTES_LARGE = Unpooled.unreleasableBuffer(
-            Unpooled.wrappedBuffer(BYTES_LARGE)).asReadOnly();
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/test/java/io/netty5/handler/codec/compression/AbstractDecoderTest.java:37]: it doesnt really matter imho

File: codec/src/test/java/io/netty5/handler/codec/compression/AbstractDecoderTest.java [hunk 11]
```diff
@@ -66,87 +60,76 @@ public void destroyChannel() {
...
+        }
     }
 
-    protected void testDecompression(final ByteBuf expected, final ByteBuf data) throws Exception {
-        assertTrue(channel.writeInbound(data));
-
-        ByteBuf decompressed = readDecompressed(channel);
-        assertEquals(expected, decompressed);
+    protected void testDecompression(final Buffer expected, final Buffer data) throws Exception {
+        assertTrue(channel.writeInbound(data.readSplit(data.readableBytes())));
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/test/java/io/netty5/handler/codec/compression/AbstractDecoderTest.java:100]: yes as the data is closed automatically later on

File: codec/src/test/java/io/netty5/handler/codec/compression/AbstractIntegrationTest.java [hunk 12]
```diff
@@ -152,35 +138,29 @@ public void testSequential() throws Exception {
...
+                }
             }
-            in.readerIndex(0);
-            assertEquals(in, decompressed);
         } finally {
-            compressed.release();
-            decompressed.release();
-            in.release();
-            closeChannels();
+            //closeChannels();
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/test/java/io/netty5/handler/codec/compression/AbstractIntegrationTest.java:163]: good catch!

File: codec/src/test/java/io/netty5/handler/codec/compression/CompressionTestUtils.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                }
+                if (msg.writableBytes() > 0) {
+                    // We also can't compose buffers that will have writer-offset gaps.
+                    // Trim off the excess with split.
+                    compositeBuffer.extendWith(msg.split().send());
+                } else {
+                    compositeBuffer.extendWith(msg.send());
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/test/java/io/netty5/handler/codec/compression/CompressionTestUtils.java:76]: sure can do this... that said it not really matters for tests imho

File: codec/src/main/java/io/netty5/handler/codec/compression/BrotliDecompressor.java [hunk 14]
```diff
@@ -86,15 +85,15 @@ private BrotliDecompressor(int inputBufferSize) throws IOException {
...
     }
 
-    private ByteBuf pull(ByteBufAllocator alloc) {
+    private Buffer pull(BufferAllocator alloc) {
         ByteBuffer nativeBuffer = decoder.pull();
         // nativeBuffer actually wraps brotli's internal buffer so we need to copy its content
-        ByteBuf copy = alloc.buffer(nativeBuffer.remaining());
+        Buffer copy = alloc.allocate(nativeBuffer.remaining());
         copy.writeBytes(nativeBuffer);
         return copy;
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/BrotliDecompressor.java:93]: lets change this.

File: codec/src/main/java/io/netty5/handler/codec/compression/FastLzCompressor.java [hunk 15]
```diff
@@ -135,30 +134,32 @@ public static Supplier<FastLzCompressor> newFactory(int level, Checksum checksum
...
                 throw new IllegalStateException();
         }
     }
 
-    private ByteBuf compressData(ByteBuf in, ByteBufAllocator allocator) {
+    private Buffer compressData(Buffer in, BufferAllocator allocator) {
         final ByteBufChecksum checksum = this.checksum;
-        ByteBuf out = allocator.buffer();
+        // TODO: Is this a good size ?
+        Buffer out = allocator.allocate(in.readableBytes() / 2);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/FastLzCompressor.java:153]: lets use `/ 1.5`

File: codec/src/main/java/io/netty5/handler/codec/compression/Lz4Decompressor.java [hunk 16]
```diff
@@ -240,44 +251,53 @@ public ByteBuf decompress(ByteBuf in, ByteBufAllocator allocator) throws Decompr
...
+                                assert compressedLength == decompressedLength;
+                                uncompressed = in.readSplit(decompressedLength);
                                 break;
                             case BLOCK_TYPE_COMPRESSED:
-                                uncompressed = allocator.buffer(decompressedLength, decompressedLength);
-
-                                decompressor.decompress(CompressionUtil.safeNioBuffer(in),
-                                        uncompressed.internalNioBuffer(
-                                                uncompressed.writerIndex(), decompressedLength));
+                                uncompressed = allocator.allocate(decompressedLength);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/Lz4Decompressor.java:264]: yeah doesn't make any difference imho. Lets leave it as it is

File: codec/src/main/java/io/netty5/handler/codec/compression/LzfDecompressor.java [hunk 17]
```diff
@@ -200,53 +199,40 @@ public ByteBuf decompress(ByteBuf in, ByteBufAllocator allocator) throws Decompr
...
-                            ByteBuf data = uncompressed;
-                            uncompressed = null;
-                            return data;
-                        } finally {
-                            if (uncompressed != null) {
-                                uncompressed.release();
-                            }
-                            if (!in.hasArray()) {
-                                recycler.releaseInputBuffer(inputArray);
+                        if (in.countReadableComponents() == 1) {
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/LzfDecompressor.java:202]: good catch..

[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/LzfDecompressor.java:202]: fixed

File: codec/src/main/java/io/netty5/handler/codec/compression/Snappy.java [hunk 18]
```diff
@@ -460,37 +468,41 @@ static int decodeLiteral(byte tag, ByteBuf in, ByteBuf out) {
...
+        out.readerOffset(readerIndex);
 
         return length;
     }
 
+    private static void copyRegion(Buffer out, int initialIndex, int offset, int length) {
+        out.readerOffset(initialIndex - offset);
+        out.copyInto(out.readerOffset(), out, out.writerOffset(), length);
+        out.skipReadable(length);
+        out.skipWritable(length);
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/Snappy.java:503]: does it matter ? But sure why not.

**Comment by normanmaurer** [comment]
Something is wrong with gzip handling ... investigating

File: codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/compression/DeflateEncoder.java [hunk 19]
```diff
@@ -131,40 +127,50 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
...
             }
-            fullCompressedContent.addComponent(true, partCompressedContent);
+            if (partCompressedContent.readerOffset() != 0) {
+                // We can't compose buffers that will have reader-offset gaps.
+                partCompressedContent.readSplit(0).close(); // Trim off already-read bytes at the beginning.
+            }
+            if (partCompressedContent.writableBytes() > 0) {
+                // We also can't compose buffers that will have writer-offset gaps.
+                // Trim off the excess with split.
+                bufferList.add(partCompressedContent.split().send());
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/compression/DeflateEncoder.java:155]: Same here. The remainder of `partCompressedContent` is not closed.

File: codec/src/main/java/io/netty5/handler/codec/compression/ZlibCompressor.java [hunk 20]
```diff
@@ -263,19 +263,19 @@ public ByteBuf finish(ByteBufAllocator allocator) {
...
-                        footer.writeByte(crcValue >>> 16);
-                        footer.writeByte(crcValue >>> 24);
-                        footer.writeByte(uncBytes);
-                        footer.writeByte(uncBytes >>> 8);
-                        footer.writeByte(uncBytes >>> 16);
-                        footer.writeByte(uncBytes >>> 24);
+                        footer.writeByte((byte) crcValue);
+                        footer.writeByte((byte) (crcValue >>> 8));
+                        footer.writeByte((byte) (crcValue >>> 16));
+                        footer.writeByte((byte) ((crcValue >>> 24)));
```
Comment (chrisvest) [inline codec/src/main/java/io/netty5/handler/codec/compression/ZlibCompressor.java:269]: Looks like there are some extra parenthesis.

**Review by chrisvest — COMMENTED** [review]
Looking good. Had a couple of more comments.

File: codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/compression/DeflateDecoder.java [hunk 21]
```diff
@@ -119,24 +118,35 @@ private Buffer decompressContent(ChannelHandlerContext ctx, WebSocketFrame msg)
...
-            compositeDecompressedContent.extendWith(extractOrCopy(ctx.bufferAllocator(),
-                    partUncompressedContent).send());
+            if (partUncompressedContent.readerOffset() != 0) {
+                // We can't compose buffers that will have reader-offset gaps.
+                partUncompressedContent.readSplit(0).close(); // Trim off already-read bytes at the beginning.
+            }
+            if (partUncompressedContent.writableBytes() > 0) {
+                // We also can't compose buffers that will have writer-offset gaps.
+                // Trim off the excess with split.
+                bufferList.add(partUncompressedContent.split().send());
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec-http/src/main/java/io/netty5/handler/codec/http/websocketx/extensions/compression/DeflateDecoder.java:143]: @chrisvest let me look into the new method and fix it inline for now

File: codec/src/main/java/io/netty5/handler/codec/compression/ZlibDecompressor.java [hunk 22]
```diff
@@ -532,17 +538,21 @@ private static boolean looksLikeZlib(short cmf_flg) {
...
             }
 
-            return allocator.buffer(Math.min(preferredSize, maxAllocation), maxAllocation);
+            Buffer buf = allocator.allocate(Math.min(preferredSize, maxAllocation));
+            buf.implicitCapacityLimit(maxAllocation);
+            return buf;
         }
 
+        // TODO: Fix me
+        /*
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply codec/src/main/java/io/netty5/handler/codec/compression/ZlibDecompressor.java]: I fixed it.... its a bit more "complicated" then before as ensureWritable(...) not throws if the implicitCapacity is hit. I think it maybe should @chrisvest WDYT ?

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
@normanmaurer I'm merging this. We can do cleanups based on #12439 in a separate PR.

**Pull request merged**